src/middlewared/middlewared/plugins/acm.py

```diff
@@ -0,0 +1,736 @@
+        names.extend(data['san'])
+        return names
+
+    @accepts(
+        Dict(
+            'acme_create',
+            Bool('tos', default=False),
+            Int('csr_id', required=True),
+            Str('directory_uri', required=True),
+            Str('name', required=True),
+            Dict('domain_dns_mapping', additional_attrs=True, required=True)
+        )
+    )
+    @job(lock='acme_cert_create')
+    def do_create(self, job, data):
+        verrors = ValidationErrors()
+
+        csr_data = self.middleware.call_sync(
+            'certificate.query',
+            [['id', '=', data['csr_id']]]
+        )
+
+        if self.middleware.call_sync(
+            'certificate.query',
+            [['name', '=', data['name']]]
+        ):
+            verrors.add(
+                'acme_create.name',
+                'A Certificate with this name already exists'
+            )
+
+        if not csr_data:
+            verrors.add(
+                'acme_create.csr_id',
+                'Specified CSR does not exist on FreeNAS system'
+            )
+        elif not csr_data[0]['CSR']:
+            verrors.add(
+                'acme_create.csr_id',
+                'Please provide a valid CSR id'
+            )
+        else:
+            csr_data = csr_data[0]
+
+        if verrors:
+            raise verrors
+
+        job.set_progress(10, 'Initial validation complete')
+
+        final_order = self.issue_certificate(job, 25, data, csr_data)
+
+        job.set_progress(95, 'Final order received from ACME server')
+
+        cert_dict = {
+            'acme': self.middleware.call_sync(
+                        'acme.registration.query',
+                        [['directory', '=', data['directory_uri']]]
+                    )[0]['id'],
+            'acme_uri': final_order.uri,
+            'certificate': final_order.fullchain_pem,
+            'CSR': csr_data['CSR'],
+            'privatekey': csr_data['privatekey'],
+            'name': data['name'],
+            'expire': final_order.body.expires.astimezone(pytz.utc).strftime("%Y-%m-%d"),
+            'chain': True if len(RE_CERTIFICATE.findall(final_order.fullchain_pem)) > 1 else False,
+            'type': CERT_TYPE_EXISTING,
+            'domain_authenticators': data['domain_dns_mapping']
+        }
+
+        for key, value in (self.middleware.call_sync(
+            'certificate.load_certificate', final_order.fullchain_pem
+        )).items():
+            cert_dict[key] = value
+
+        # save cert and other useful attributes
+        cert_id = self.middleware.call_sync(
+            'datastore.insert',
+            self._config.datastore,
+            cert_dict, {
+                'prefix': self._config.datastore_prefix
+            }
+        )
+
+        self.middleware.call_sync(
+            'service.start',
+            'ix-ssl',
+            {'onetime': False}
+        )
+
+        return self.middleware.call_sync(
+            'acme.query',
```

Comment:
`acme._get_instance` could work here in call_sync

Reply:
I have a todo for that : P